### PR TITLE
Support lossless serialization for Git dependencies in lockfile

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR adds lossless deserialization for `GitSourceDist` distributions in the lockfile. Specifically, we now properly preserve the requested revision, the subdirectory, and the precise Git commit SHA.

## Test Plan

`cargo test`

### CODE REVIEW

File: crates/uv-resolver/src/lock.rs [hunk 1]
```diff
@@ -716,6 +772,54 @@ impl SourceDist {
...
+        GitReference::ShortCommit(rev)
+        | GitReference::BranchOrTag(rev)
+        | GitReference::BranchOrTagOrCommit(rev)
+        | GitReference::NamedRef(rev)
+        | GitReference::FullCommit(rev) => {
+            url.query_pairs_mut()
+                .append_pair("rev", rev.to_string().as_str());
+        }
+        GitReference::DefaultBranch => {}
+    }
```
Comment (charliermarsh) [inline crates/uv-resolver/src/lock.rs:808]: I feel like these need to be unified (part of what I was getting at in previous comments).

Comment (charliermarsh) [reply crates/uv-resolver/src/lock.rs:808]: Perhaps `GitReference` becomes internal to the crate, and that crate uses `GitSource` as its public interface? Not sure. It's useful (within the crate) to know that a ref can't be a commit, for example.

File: crates/uv-resolver/src/lock.rs [hunk 2]
```diff
@@ -581,18 +631,24 @@ impl GitSource {
...
         for (key, val) in url.query_pairs() {
-            kind = match &*key {
-                "tag" => GitSourceKind::Tag(val.into_owned()),
-                "branch" => GitSourceKind::Branch(val.into_owned()),
-                "rev" => GitSourceKind::Rev(val.into_owned()),
+            match &*key {
+                "tag" => kind = GitSourceKind::Tag(val.into_owned()),
+                "branch" => kind = GitSourceKind::Branch(val.into_owned()),
+                "rev" => kind = GitSourceKind::Rev(val.into_owned()),
+                "subdirectory" => subdirectory = Some(val.into_owned()),
```
Comment (charliermarsh) [inline crates/uv-resolver/src/lock.rs:640]: I'm also storing the subdirectory in the query string.

File: crates/uv-resolver/src/lock.rs [hunk 3]
```diff
@@ -279,6 +282,33 @@ impl Distribution {
...
+                    );
+                    if let Some(precise) = &git.precise {
+                        if let Ok(precise) = GitSha::from_str(precise) {
+                            git_url.set_precise(precise);
+                        }
+                    }
+
+                    // Reconstruct the PEP 508-compatible URL from the `GitSource`.
+                    // TODO(charlie): This shouldn't be necessary; it's only necessary because we're
+                    // still losing the `GitUrl` somewhere along the path to installation.
```
Comment (charliermarsh) [inline crates/uv-resolver/src/lock.rs:299]: Not actually sure _what_ we should track here. The verbatim URL?

**Comment by charliermarsh** [comment]
I would be fine merging as-is but I expect the abstractions to evolve a bit as I add similar logic for direct URLs.

File: crates/uv-resolver/src/lock.rs [hunk 4]
```diff
@@ -716,6 +756,80 @@ impl SourceDist {
...
+
+    // Put the precise commit in the fragment.
+    url.set_fragment(
+        git_dist
+            .git
+            .precise()
+            .as_ref()
+            .map(GitSha::to_string)
+            .as_deref(),
+    );
```
Comment (BurntSushi) [inline crates/uv-resolver/src/lock.rs:828]: I think it should be an error if we don't have a precise commit right? Without the precise commit revision, the lock file cannot guarantee reproducibility.

File: crates/uv-resolver/src/lock.rs (hunk 1 above)
Comment (BurntSushi) [reply crates/uv-resolver/src/lock.rs:818]: Can you say more about what unification means in this context? I think as long as we require that a specific revision appears in the lock file (and the sub-directory if it exists), then I think that's all we actually need.

File: crates/uv/tests/lock.rs [hunk 5]
```diff
@@ -212,10 +212,10 @@ fn lock_git() -> Result<()> {
     [[distribution]]
     name = "anyio"
     version = "3.7.0"
-    source = "git+https://github.com/agronholm/anyio"
+    source = "git+https://github.com/agronholm/anyio?rev=3.7.0#f7a880ffac4766efb39e6fb60fc28d944f5d2f65"
```
[the remark this answers is not in the document]
Comment (BurntSushi) [reply crates/uv/tests/lock.rs:215]: Another thought here is that tags/branches/whatever are mutable. So if we have `rev=3.7.0` but it doesn't actually match the precision revision, then installation could error or emit a warning. But I'm not 100% certain that's the right UX. (And I could see arguments saying that's actually the wrong UX.)

**Review by BurntSushi — APPROVED** [review]
Happy to have this come in as-is, with maybe a comment or two noting FIXMEs. :)

File: crates/uv-resolver/src/lock.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv-resolver/src/lock.rs:818]: Mostly just commenting that `GitReference` and `GitSource` are _nearly_ identical, except that `GitReference` contains more variants that break down `GitSource::Rev` in a way that's purely based on analyzing the value.

**Comment by charliermarsh** [comment]
I'll fix up the "require a SHA" part in this PR.

**Pull request merged**